conf_file.cpp

```diff
@@ -16,268 +16,416 @@
-    while (section) {
-        if (strcaseeq(section->name, entry)) {
-            _current_section = section;
-            free(entry);
-            return 0;
-        }
-        section = section->next;
+    if (equal_pos == (entry + entry_len - 1)) {
+        log_error("On file %s: Line %d: Missing value on config", filename, line);
+        return -EINVAL;
     }
 
-    // Section is new.
-    section = (struct section *)calloc(1, sizeof(struct section));
-    assert_or_return(section, -ENOMEM);
+    key = entry;
+    key_len = equal_pos - entry;
+    value = equal_pos + 1;
+    value_len = entry_len - key_len - 1;
 
-    section->name = entry;
-    section->next = _sections;
-    _sections = section;
+    _trim(&key, &key_len);
+    _trim(&value, &value_len);
 
-    _current_section = section;
+    c = _find_config(s, key, key_len);
+    if (c) {
+        c->value = value;
+        c->value_len = value_len;
+    } else
+        s->configs = c = new config{key, value, key_len, value_len, s->configs};
 
     return 0;
 }
 
-int ConfFile::_add_config(char *entry, int line)
+void ConfFile::_trim(const char **str, size_t *len)
 {
-    char *equal_pos;
-    struct config *config;
+    const char *s = *str;
+    const char *end = s + *len;
 
-    if (!(equal_pos = strchr(entry, '='))) {
-        log_error("On file %s: Line %d: Missing '=' on config", _filename, line);
-        return -EINVAL;
+    while (isspace(*s) && s < end)
+        s++;
+
+    while (end > s && isspace(*(end - 1)))
+        end--;
+
+    *len = end - s;
+    *str = s;
+}
+
+ConfFile &ConfFile::global_instance()
+{
+    static ConfFile conf;
+
+    return conf;
+}
+
+void ConfFile::release_all()
+{
+    struct conffile *f;
+    struct section *s;
+    struct config *c;
+
+    while (_files) {
+        f = _files;
+        _files = _files->next;
+
+        munmap(f->addr, f->len);
+        close(f->fd);
+        delete f;
     }
 
-    if (equal_pos == entry) {
-        log_error("On file %s: Line %d: Missing name on config", _filename, line);
-        return -EINVAL;
+    while (_sections) {
+        s = _sections;
+        _sections = _sections->next;
+        while (s->configs) {
+            c = s->configs;
+            s->configs = s->configs->next;
+            delete c;
+        }
+
+        delete s;
     }
+}
 
-    if (equal_pos == (entry + strlen(entry) - 1)) {
-        log_error("On file %s: Line %d: Missing value on config", _filename, line);
-        return -EINVAL;
+int ConfFile::_extract_options_from_section(struct section *s, const OptionsTable table[],
+                                            size_t table_len, void *data)
+{
+    struct config *c;
+    int ret;
+    size_t i;
+    void *storage;
+
+    for (i = 0; i < table_len; i++) {
+        c = _find_config(s, table[i].key, strlen(table[i].key));
+        if (!c) {
+            if (table[i].required) {
+                log_error("Required field '%s' not found in section '%.*s'", table[i].key,
```

Comment:
Yeah, I've had an error here and had to grep to find which file caused the error. It's more of a nice to have... not sure if someone would have a long chain of overriding configs...